presto-orc/src/main/java/com/facebook/presto/orc/stream/BooleanInputStream.java

```diff
@@ -132,13 +133,64 @@ public int getSetBits(int batchSize, boolean[] vector)
             throws IOException
     {
         int count = 0;
-        for (int i = 0; i < batchSize; i++) {
+        int bitsInFirstByte = Math.min(batchSize, bitsInData);
+        for (int i = 0; i < bitsInFirstByte; i++) {
+            vector[i] = nextBit();
+            count += vector[i] ? 1 : 0;
+        }
+        verify(bitsInData == 0);
+        int wholeBytes = (batchSize - bitsInFirstByte) / 8;
+        int numFilled = bitsInFirstByte;
+        for (int i = 0; i < wholeBytes; i++) {
+            byte data = byteStream.next();
+            count += Integer.bitCount(data);
+            vector[numFilled] = (data & 0x80) != 0;
+            vector[numFilled + 1] = (data & 0x40) != 0;
+            vector[numFilled + 2] = (data & 0x20) != 0;
+            vector[numFilled + 3] = (data & 0x10) != 0;
+            vector[numFilled + 4] = (data & 0x08) != 0;
+            vector[numFilled + 5] = (data & 0x04) != 0;
+            vector[numFilled + 6] = (data & 0x02) != 0;
+            vector[numFilled + 7] = (data & 0x01) != 0;
+            numFilled += 8;
+        }
+        for (int i = numFilled; i < batchSize; i++) {
             vector[i] = nextBit();
             count += vector[i] ? 1 : 0;
         }
         return count;
     }
 
+    // vector[i] is set to the bit at offsets[i] - offsetBase for i
+    // from 0 to numOffsets. The offset[i] - offsetBase values are
+    // relative to the current position of the stream. 0 would mean
+    // the value returned by a call to next(). The stream is advanced
+    // by numBits. The offset values are increasing and non-repeating.
+    public void getSetBits(int[] offsets, int numOffsets, int offsetBase, int numBits, boolean[] vector)
+            throws IOException
+    {
+        if (offsets[0] == offsetBase && offsets[numOffsets - 1] == offsetBase + numOffsets - 1) {
+            getSetBits(numOffsets, vector);
+            if (numBits > offsets[numOffsets - offsetBase - 1] + 1) {
```

Comment:
How does this calculation work? It seems strange to mix numOffsets and offsetBase in a single calculation. Can't this be simplified to:

```
if (numBits > numOffsets) {
    skip(numBits - numOffsets);
}
```